src/variableSubstitution.ts

```diff
@@ -0,0 +1,154 @@
+import * as path from "path";
+import { TextDocument, window, workspace, WorkspaceFolder } from "vscode";
+import { XMLFileAssociation } from "./extension";
+
+/**
+ * Represents a variable that refers to a value that can be resolved
+ */
+class VariableSubstitution {
+
+  private varName: string;
+  private varKind: VariableSubstitutionKind;
+  private getValue: (currentFileUri: string, currentWorkspaceUri: string) => string;
+  private replaceRegExp: RegExp | undefined;
+
+  /**
+   *
+   * @param name the name of this variable
+   * @param kind the kind of this variable
+   * @param getValue a function that resolves the value of the variable
+   */
+  constructor(name: string, kind: VariableSubstitutionKind, getValue: (currentFileUri: string, currentWorkspaceUri: string) => string) {
+    this.varName = name;
+    this.varKind = kind;
+    this.getValue = getValue;
+  }
+
+  public get name(): string {
+    return this.varName;
+  }
+
+  public get kind(): VariableSubstitutionKind {
+    return this.varKind;
+  }
+
+  /**
+   * Returns the string with the references to this variable replaced with the value, or the original string if the value cannot be resolved
+   *
+   * @param original the string to substitute variable value
+   * @param currentFileUri the uri of the currently focused file, as a string
+   * @param currentWorkspaceUri the uri of the root of the currently open workspace, as a string
+   * @returns the string with the references to this variable replaced with the value, or the original string if the value cannot be resolved
+   */
+  public substituteString(original: string, currentFileUri: string, currentWorkspaceUri: string): string {
+    const value: string = this.getValue(currentFileUri, currentWorkspaceUri);
+    return value ? original.replace(this.getReplaceRegExp(), value) : original;
+  }
+
+  /**
+   * Returns a regex that matches all references to the variable
+   *
+   * Lazily initialized
+   *
+   * @returns a regex that matches all references to the variable
+   */
+  public getReplaceRegExp(): RegExp {
+    if (!this.replaceRegExp) {
+      this.replaceRegExp = new RegExp('\\$\\{' + `${this.name}` + '\\}', 'g');
+    }
+    return this.replaceRegExp;
+  }
+}
+
+enum VariableSubstitutionKind {
+  Workspace,
+  File
+}
+
+const SETTINGS_VARIABLES: VariableSubstitution[] = [
```

Comment:
Very good job! Could you:

 * rename SETTINGS_VARIABLES to VARIABLE_SUBSTITIONS
 * move this const declaration on the top of the file
 * add a comment before the const to explain that of we want to add a new variable substitution, just update this const array.

Reply:
I get a linter error when I try to declare VARIABLE_SUBSTITUTIONS before VariableSubstitution, but I made the other changes.